Include/internal/pycore_frame.h

```diff
@@ -23,12 +23,76 @@ struct _frame {
     char f_trace_opcodes;       /* Emit per-opcode trace events? */
     char f_fast_as_locals;      /* Have the fast locals of this frame been converted to a dict? */
     /* The frame data, if this frame object owns the frame */
-    PyObject *_f_frame_data[1];
+    char *_f_frame_data[1];
 };
 
 extern PyFrameObject* _PyFrame_New_NoTrack(PyCodeObject *code);
 
 
+/* Tagged pointers on the stack */
+
+typedef union {
+    uintptr_t val;
+    PyObject *obj;
+} _tagged_ptr;
+
+static_assert(sizeof(_tagged_ptr) == sizeof(PyObject *),
+              "tagged pointer must be the same size as object pointer");
+
+static inline bool is_tagged(_tagged_ptr tp) {
+    return tp.val & 1;
+}
+
+static inline PyObject *detag(_tagged_ptr tp) {
+    return (PyObject *) (tp.val & ~1);
+}
+
+static inline _tagged_ptr untagged(PyObject *p) {
+    return (_tagged_ptr) { .obj = p };
+}
+
+static inline _tagged_ptr tagged(PyObject *p) {
+    return (_tagged_ptr) { .val = (uintptr_t) p | 1 };
+}
+
+static inline void decref_unless_tagged(_tagged_ptr tp) {
+    if (!is_tagged(tp)) {
+        Py_DECREF(detag(tp));
+    }
+}
+
+static inline void xdecref_unless_tagged(_tagged_ptr tp) {
+    if (!is_tagged(tp)) {
+        Py_XDECREF(detag(tp));
+    }
+}
+
+static inline void incref_if_tagged(_tagged_ptr tp) {
+    if (is_tagged(tp)) {
+        Py_INCREF(detag(tp));
+    }
```

Comment:
Have you seen the branch-free alternative to this I [proposed here](https://github.com/faster-cpython/ideas/issues/443#issuecomment-1209736091)? 

The primitive for increffing is branch-free and the one to decref does a pointer comparison rather than going through a struct.

Reply:
The line
```
untagged_ob->ob_refcnt += ((uintptr_t)ob) & 1ull;
```
in your proposal would end up always doing a memory read and write, which is exactly what we're trying to avoid here.

Reply:
I trust branch prediction more than memory latency. :-)